Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -253,3 +254,456 @@ theorem Language.IsContextFree.reverse (L : Language T) :
   fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
 
 end closure_reversal
+
+section lift_sink
+
+/- This section contains only auxilliary constructions that will shorten upcoming proofs of
+closure properties. When combining several grammars together, we usually want to take a sum type of
+their nonterminal types and lift respective nonterminals to this sum type. We subsequently show
+that the resulting grammar preserves derivations of those strings that may contain any terminals but
+only the proper nonterminals. The lifting operation must be injective. The sinking operation must be
+injective on those symbols where it is defined. -/
+
+/-- Lifting `Symbol` to a larger nonterminal type. -/
+def liftSymbol {N₀ N : Type*} (liftN : N₀ → N) : Symbol T N₀ → Symbol T N
+  | Symbol.terminal t => Symbol.terminal t
+  | Symbol.nonterminal n => Symbol.nonterminal (liftN n)
+
+/-- Sinking `Symbol` from a larger nonterminal type; may return `none`. -/
+def sinkSymbol {N₀ N : Type*} (sinkN : N → Option N₀) : Symbol T N → Option (Symbol T N₀)
```

Comment:
and here, following `List`, this should be `Symbol.filterMap`, or alternatively `Symbol.reduceOption ∘ Symbol.map sinkN`

Reply:
I am a bit worried that `Symbol.map` and `Symbol.filterMap` will make it look like they are both intended to be used in the same direction. The intention here is that they are used as almost inverses.